Support cartesian_bounds aggregation on point and shape

craigtaverner opened the pull request:
Fixes https://github.com/elastic/elasticsearch/issues/90157

craigtaverner committed:
Support cartesian_bounds aggregation on point and shape

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner committed:
Update docs/changelog/91298.yaml

elasticsearchmachine commented:
Hi @craigtaverner, I've updated the changelog YAML for you.

craigtaverner committed:
Update docs/changelog/91298.yaml

craigtaverner committed:
Fixed cartesian parse exceptions to mimic previous behavior

craigtaverner committed:
Re-added removed interfaces to fix failing tests

craigtaverner committed:
Generalized AbstractGeoTestCase for use in Cartesian tests

And added new CartesianBoundsIT and CartesianCentroidIT

craigtaverner committed:
Fixed bug in encoding missing cartesian points

craigtaverner committed:
Simplify cartesian bounds

* Removing the neg/pos left/right split used in bounds because this is only relevant to GeoBounds
* Revert GeoBounds to no longer share common code with CartesianBounds (including aggregators)
* Make Point and Shape aggregators share common code

craigtaverner committed:
Fixed geo comment

iverase reviewed (changes requested):
I left some minor comments specially for the visibility of `Shapes.ShapeValue`.
Two things:

1) You have some weird file named `jffi4926421121997545579.dylib` which seems some left over.

~~2) We need to check if it is ok to change the API of GeoBounds, not sure if that is an. API we cannot change in a minor.~~
I check things again and everything is clear.

iverase commented on server/src/internalClusterTest/java/org/elasticsearch/search/aggregations/metrics/GeoBoundsIT.java:
```diff
@@ -43,15 +44,15 @@ public void testSingleValuedField() throws Exception {
         assertSearchResponse(response);
 
-        GeoBounds geoBounds = response.getAggregations().get(aggName);
+        SpatialBounds<GeoPoint> geoBounds = response.getAggregations().get(aggName);
```
Can be use here GeoBounds instead?

craigtaverner replied:
I was in the process of refactoring. I'll finish the job and then you can decide.

iverase commented on server/src/internalClusterTest/java/org/elasticsearch/search/aggregations/metrics/GeoBoundsIT.java:
```diff
@@ -69,25 +70,28 @@ public void testSingleValuedField_getProperty() throws Exception {
         assertThat(global.getAggregations().asMap().size(), equalTo(1));
 
-        GeoBounds geobounds = global.getAggregations().get(aggName);
+        SpatialBounds<GeoPoint> geobounds = global.getAggregations().get(aggName);
```
The same here

craigtaverner replied:
I've done this for the two dateline tests, but all the rest moved into shared code.

craigtaverner committed:
Delete incorrectly added file

craigtaverner commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/metrics/CartesianBoundsAggregationBuilder.java:
```diff
@@ -0,0 +1,111 @@
+    @Override
+    public Version getMinimalSupportedVersion() {
+        // TODO: Should we have a version here? We only release this for 8.5 or 8.6
+        return Version.V_EMPTY;
```
I noticed that none of the other aggregations had values set here, even when recently added.

craigtaverner replied:
OK. I spoke too soon, while the first 4 I looked at did not specify anything, a more careful analysis shows that 4 of the 13 known Geo aggregations did have settings:

* `GeoHexGridAggregationBuilder` -> `Version.V_8_1_0`
* `GeoLineAggregationBuilder` -> `Version.V_7_11_0`
* `GeoGridQueryBuilder` -> `Version.V_8_3_0`
* `GeoTileGridAggregationBuilder` -> `Version.V_7_0_0`

So I will add `Version.V_8_6_0` to the current one.

craigtaverner commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/metrics/CartesianShapeBoundsAggregator.java:
```diff
@@ -0,0 +1,58 @@
+            public void collect(int doc, long bucket) throws IOException {
+                if (values.advanceExact(doc)) {
+                    maybeResize(bucket);
+                    ShapeValues.ShapeValue value = values.value();
```
Done

craigtaverner committed:
Set correct minimum version for new cartesian aggs

craigtaverner committed:
Fixed specific types for ShapeValues

craigtaverner committed:
Completed Centroid/Bounds aggregation integration tests refactoring

An earlier commit started generalizing Geo/Cartesian integration tests for
centroid and bounds aggregations, but did not complete the job.

This commit completes the work, since almost all tests are identical between
cartesian and geo, except tests related to the dateline.

iverase commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialPlugin.java:
```diff
@@ -172,7 +176,15 @@ public List<AggregationSpec> getAggregations() {
+                usage.track(
+                    SpatialStatsAction.Item.CARTESIANBOUNDS,
+                    CartesianBoundsAggregationBuilder.PARSER    // TODO: Should we have a license for this?
+                )
```
No, license for CartesianBounds should be the same as for GeoBounds, so it is basic.

craigtaverner committed:
Removed licensing comment

craigtaverner committed:
GeoBoundsIT expects wrapLongitude(false)

craigtaverner committed:
Fixed CartesianBoundsIT flakiness and removed geo-specific code

craigtaverner committed:
Reduced CartesianCentroidIT flakiness by normalizing to floats

iverase reviewed (approved):
LGTM

elasticsearchmachine commented:
@craigtaverner according to this PR's labels, I need to update the changelog YAML, but I can't because the PR is closed. Please either update the changelog yourself on the appropriate branch, or adjust the labels. Specifically:

   * The PR is labelled `release highlight` but the changelog has no `highlight` section
